### ISSUE TITLE
Support large data inputs and DDP consumption

### ISSUE THREAD

**Issue by henripal** [pytorch/data#245]
### 🚀 The feature

DataPipe support for very large datasets at source and sink levels:

*Source level*: 
Support for large sharded datasets (think 100M protein sequences across 1000s of CSV files, 100M molecular graphs in 1000s of pickled binaries, ...)
- allow for loading and batched pre-processing
- some support for shuffling

*Sink level*: 
Support for easy parallelization of dataloading when training with `DistributedDataParallel`
- distribute the input data across GPUs and workers without custom code
- provide options to deal with uneven data distribution at the dataset level

### Motivation, pitch

Writing training pipelines for very large datasets is currently quite labor-intensive, but often uses several concepts that I think could be abstracted away.
1 - pre-processing steps are often vectorize-able, so loading the data like in the README example:

```python
@functional_datapipe("parse_csv_files")
class CSVParserIterDataPipe(IterDataPipe):
    def __init__(self, dp, **fmtparams):
        self.dp = dp
        self.fmtparams = fmtparams

    def __iter__(self):
        for filename, stream in self.dp:
            reader = csv.reader(stream, **self.fmtparams)
            for row in reader:
                yield filename, row
```

would lead to inefficiencies. A proposed way would be some implementation that would allow batching and batch processing of datasets.

2 - when using DDP with datasets above, in the setting in which there are multiple GPUs and multiple workers processing and loading the data per GPU, I think the current approach is to:
- shard the initial dataset (in `n` files, for example)
- distribute the files amongst the workers*GPUs to ensure that every GPU gets a different slice of the dataset

This could be facilitated if there was a way to load a dataset in chunks and distribute the chunks across workers automatically using DataPipes.

### Alternatives

Some in progress/parallel implementations of this I've come across:
- An API was proposed and implemented in https://github.com/pytorch/pytorch/pull/26545
- Petastorm project tries to address this problem specifically for parquet files: https://github.com/uber/petastorm - however found this to not work in practice because does not guarantee even shards: https://github.com/uber/petastorm/issues/508
- YogaDL is an open source data sharding that adds a caching layer for initial dataset shuffling: https://github.com/determined-ai/yogadl

**Comment by ejguan** [comment]
TorchData is introduced for loading large dataset with better syntax.
For you first pitch, I do not fully understand why it is not efficient. We provide `batch` for you to batch prior data to a list. Then, you can provide vectorized function to apply on a batch.
The example you point out is basically reading rows from a single csv file. I believe you want to yield multiple rows at the same time. But, in terms of the efficiency, I don't see any difference because it's not vectorization.

For the second pitch, we are working on dynamic sharding for multiprocessing and distributed training. Then, you don't need to manually provide hard-coded number of shards within the implementation of your dataset. We will let you know when the feature is ready.

**Comment by henripal** [comment]
Thank you! Let us know if we can help test or give feedback in any way!

**Comment by josiahls** [comment]
Not sure if this is a similar thing, but I'm concerned about being able to cache certain points in a pipeline (maybe there already is something?).

My problem is simple: split a datapipe into train/valid datapipes.

For example, I want to split a datapipe into train/valid datapipes. I think this would be a combo of `fork` and `filter`. The problem is that per iter, the filter may have to iterate through 10000+ train elements get to a validation element and vice versa.

I've been messing around with a `Cacher` data pipe that could help with storing the train and valid. If this isn't the correct method of avoiding this, I'd appreciate an example to be pointed to.

**Comment by ejguan** [comment]
@josiahls 
You can use `Demux` https://github.com/pytorch/pytorch/blob/1d497114e7e13822df403082975d4b212d836207/torch/utils/data/datapipes/iter/combining.py#L215 for this use case.

**Comment by s-udhaya** [comment]
@ejguan Is there any update on the dynamic sharding for multiprocessing and distributed training feature?
